### DISCUSSION TITLE
This grammar causes parsing to hang indefinitely

### DISCUSSION THREAD

**Discussion by jdidion** [pest-parser/pest#797] [Q&A] [answered]
I've written a grammar for the [WDL](https://github.com/openwdl/wdl) language.

The grammar is here:

https://github.com/jdidion/pest-wdl/blob/main/src/wdl.pest

I am observing the problem when trying to parse an `apply` expression (i.e. a function call) with non-primitive arguments. The problem also happens when calling a function within a string interpolation, even without supplying any arguments.

```
version 1.1

workflow Test {
    String works = sep(" ", 1)
    String hangs = sep(" ", [])
    String also_hangs = "~{sep()}"
}
```

Is there any way to tell pest to print out its state as it's parsing so I can see where it's getting hung up?

Any help is greatly appreciated!

**Comment by tomtau** [comment]
Maybe you can try using the grammar debugger: https://github.com/pest-parser/pest/discussions/739 ?

**Comment by jdidion** [comment, marked as answer]
I'm finding that rules like the following are very slow:

```
equality = _{ not_equal | equal | comparison }
equal = { comparison ~ ( eq ~ comparison )+ }
not_equal = { comparison ~ ( ne ~ comparison )+ }
```

If I re-write this as

```
equality = { comparison ~ ( (eq | ne ) ~ comparison )* }
```

parsing time decreases by an order-of-magnitude. I am not sure if this is the root cause yet. I'm going to re-write the expression grammar and see if that resolves the issue.

**Comment by jdidion** [comment]
Yes! That seems to fix the issue. Is there any place where these anti-patterns are being collected? I'm sure others have run into this issue before.

**Reply by tomtau** [reply]
Usually in https://github.com/pest-parser/pest/discussions/categories/q-a so I converted this issue into a discussion
